## Midterm project: Equity Portfolio Management

### Data Preparation

Download the historial daily data of the entire 2018 for the 10 stocks 

```python
universe = ['IBM', 'MSFT', 'GOOG', 'AAPL', 'AMZN', 'META', 'NFLX', 'TSLA', 'ORCL', 'SAP']
```


You should have 10 csv files on your disk now. IBM.csv, MSFT.csv, etc. We call the 10 stocks "universe" which is the entire stock market you can trade.


![image.png](image.png)


### Retrieve the "Close" and "Adj Close" values for each stock

You will create a dataframe where there are 20 columns for the 10 stocks, each row is the "Close" and "Adj Close" prices for the 10 stocks on each day, in the order of the business days in 2018. Assume all buy/sell on the "Close" prices and there is no transaction cost.

In [23]:
import pandas as pd
import yfinance as yf
import plotly.graph_objects as go
from colorama import Fore, Back, Style, init

# init()
# pd.options.display.float_format = '{:,.2f}'.format

In [31]:
print([attr for attr in dir(Fore) if not attr.startswith('_')])
print([attr for attr in dir(Back) if not attr.startswith('_')])
print([attr for attr in dir(Style) if not attr.startswith('_')])

['BLACK', 'BLUE', 'CYAN', 'GREEN', 'LIGHTBLACK_EX', 'LIGHTBLUE_EX', 'LIGHTCYAN_EX', 'LIGHTGREEN_EX', 'LIGHTMAGENTA_EX', 'LIGHTRED_EX', 'LIGHTWHITE_EX', 'LIGHTYELLOW_EX', 'MAGENTA', 'RED', 'RESET', 'WHITE', 'YELLOW']
['BLACK', 'BLUE', 'CYAN', 'GREEN', 'LIGHTBLACK_EX', 'LIGHTBLUE_EX', 'LIGHTCYAN_EX', 'LIGHTGREEN_EX', 'LIGHTMAGENTA_EX', 'LIGHTRED_EX', 'LIGHTWHITE_EX', 'LIGHTYELLOW_EX', 'MAGENTA', 'RED', 'RESET', 'WHITE', 'YELLOW']
['BRIGHT', 'DIM', 'NORMAL', 'RESET_ALL']


In [2]:
# Load dataframes

df_ibm = yf.download("IBM", start="2018-01-01", end="2018-12-31", auto_adjust=False, actions=True)
df_msft = yf.download("MSFT", start="2018-01-01", end="2018-12-31", auto_adjust=False, actions=True)
df_goog = yf.download("GOOG", start="2018-01-01", end="2018-12-31", auto_adjust=False, actions=True)
df_aapl = yf.download('AAPL', start='2018-01-01', end='2018-12-31', auto_adjust=False, actions=True)
df_amzn = yf.download("AMZN", start="2018-01-01", end="2018-12-31", auto_adjust=False, actions=True)
df_meta = yf.download("META", start="2018-01-01", end="2018-12-31", auto_adjust=False, actions=True)
df_nflx = yf.download("NFLX", start="2018-01-01", end="2018-12-31", auto_adjust=False, actions=True)
df_tsla = yf.download("TSLA", start="2018-01-01", end="2018-12-31", auto_adjust=False, actions=True)
df_orcl = yf.download("ORCL", start="2018-01-01", end="2018-12-31", auto_adjust=False, actions=True)
df_sap = yf.download("SAP", start="2018-01-01", end="2018-12-31", auto_adjust=False, actions=True)

[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed


In [3]:
# Save to csv

df_ibm.to_csv("IBM_2018.csv")
df_msft.to_csv("MSFT_2018.csv")
df_goog.to_csv("GOOG_2018.csv")
df_aapl.to_csv("AAPL_2018.csv")
df_amzn.to_csv("AMZN_2018.csv")
df_meta.to_csv("META_2018.csv")
df_nflx.to_csv("NFLX_2018.csv")
df_tsla.to_csv("TSLA_2018.csv")
df_orcl.to_csv("ORCL_2018.csv")
df_sap.to_csv("SAP_2018.csv")

In [4]:
def read_stock_csv(filename):
    df = pd.read_csv(filename, header=[0,1])
    df.columns = [f"{price.strip().lower().replace(' ','_')}_{ticker.strip().lower()}" for price, ticker in df.columns]
    df = df.iloc[1:]
    df = df.rename(columns={df.columns[0]: 'Date'}) 
    df.index = range(1, len(df) + 1)

    for col in df.columns:
        if col != 'Date':
            df[col] = pd.to_numeric(df[col], errors='coerce')
    
    return df

universe = [
    'AAPL',
    'AMZN',
    'GOOG',
    'IBM',
    'META',
    'MSFT',
    'NFLX',
    'ORCL',
    'TSLA',
    'SAP',
]

stocks = {
    ticker: read_stock_csv(f"{ticker}_2018.csv") for ticker in universe
}

# for ticker, df in stocks.items():
#     print(ticker, df.shape)

# Unpacking the stocks into individual variables for easier access
(df_aapl, 
 df_amzn, 
 df_goog, 
 df_ibm,
 df_meta, 
 df_msft,
 df_nflx, 
 df_orcl, 
 df_tsla, 
 df_sap) = [stocks[t] for t in universe]

# df_aapl
df_ibm

,Date,adj_close_ibm,close_ibm,dividends_ibm,high_ibm,low_ibm,open_ibm,stock_splits_ibm,volume_ibm
1,2018-01-02,103.412827,147.466537,0.0,148.001907,146.787766,147.705551,0.0,4395815.0
2,2018-01-03,106.255432,151.520081,0.0,152.782028,149.455063,150.420654,0.0,9875914.0
3,2018-01-04,108.407478,154.588913,0.0,155.181641,152.361374,152.629059,0.0,7903785.0
4,2018-01-05,108.937103,155.344162,0.0,155.736145,154.015289,155.296371,0.0,5434807.0
5,2018-01-08,109.594124,156.281067,0.0,156.701721,154.588913,155.506699,0.0,5478425.0
...,...,...,...,...,...,...,...,...,...
246,2018-12-21,77.740990,106.061188,0.0,110.210327,105.583176,107.552582,0.0,10864279.0
247,2018-12-24,75.379478,102.839386,0.0,106.118546,102.772469,105.066925,0.0,3997184.0
248,2018-12-26,78.056328,106.491394,0.0,106.491394,101.281075,103.250481,0.0,6951507.0
249,2018-12-27,79.731125,108.776291,0.0,108.776291,104.655830,105.152962,0.0,6323698.0


In [5]:
out = []

for ticker in universe:
    df = stocks[ticker]
    t = ticker.lower()
    out.append(df[f"adj_close_{t}"])
    out.append(df[f"close_{t}"])

    # Percentage change calculations
    pct_change = (
    (df[f'adj_close_{t}'] - df[f'adj_close_{t}'].shift(5)) 
    / df[f'adj_close_{t}'].shift(5) * 100
    ).round(2)
    out.append(pct_change.rename(f'5day_pct_change_{t}'))

    # Add dividend column
    dividend_col = f"dividends_{t}"
    if dividend_col in df.columns:
        out.append(df[dividend_col])
    else:
        out.append(pd.Series(0.0, index=df.index, name=dividend_col))

df_universe = pd.concat(out, axis=1) 
df_universe.insert(0, 'Date', stocks[universe[0]]['Date'].values)
df_universe.index = range(1, len(df_universe) + 1) 

df_universe

,Date,adj_close_aapl,close_aapl,5day_pct_change_aapl,dividends_aapl,adj_close_amzn,close_amzn,5day_pct_change_amzn,dividends_amzn,adj_close_goog,...,5day_pct_change_orcl,dividends_orcl,adj_close_tsla,close_tsla,5day_pct_change_tsla,dividends_tsla,adj_close_sap,close_sap,5day_pct_change_sap,dividends_sap
1,2018-01-02,40.304176,43.064999,NaN,0.0,59.450500,59.450500,NaN,0.0,52.816380,...,NaN,0.0,21.368668,21.368668,NaN,0.0,99.060287,112.389999,NaN,0.0
2,2018-01-03,40.297150,43.057499,NaN,0.0,60.209999,60.209999,NaN,0.0,53.683270,...,NaN,0.0,21.150000,21.150000,NaN,0.0,99.871178,113.309998,NaN,0.0
3,2018-01-04,40.484341,43.257500,NaN,0.0,60.479500,60.479500,NaN,0.0,53.877666,...,NaN,0.0,20.974667,20.974667,NaN,0.0,101.404808,115.050003,NaN,0.0
4,2018-01-05,40.945274,43.750000,NaN,0.0,61.457001,61.457001,NaN,0.0,54.662720,...,NaN,0.0,21.105333,21.105333,NaN,0.0,102.532990,116.330002,NaN,0.0
5,2018-01-08,40.793186,43.587502,NaN,0.0,62.343498,62.343498,NaN,0.0,54.896305,...,NaN,0.0,22.427334,22.427334,NaN,0.0,101.184471,114.800003,NaN,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
246,2018-12-21,35.796207,37.682499,-8.91,0.0,68.872498,68.872498,-13.47,0.0,48.578178,...,-5.58,0.0,21.318001,21.318001,-12.56,0.0,86.034668,96.139999,-4.80,0.0
247,2018-12-24,34.870007,36.707500,-10.44,0.0,67.197998,67.197998,-11.63,0.0,48.413532,...,-6.65,0.0,19.692667,19.692667,-15.22,0.0,85.014496,95.000000,-4.64,0.0
248,2018-12-26,37.325607,39.292500,-5.36,0.0,73.544998,73.544998,-5.19,0.0,51.549782,...,-2.75,0.0,21.739332,21.739332,-3.25,0.0,88.039215,98.379997,-1.70,0.0
249,2018-12-27,37.083382,39.037498,-2.95,0.0,73.082001,73.082001,-2.24,0.0,51.768978,...,-3.19,0.0,21.075333,21.075333,-5.06,0.0,88.585098,98.989998,-1.12,0.0


In [6]:
pct_cols = [f"5day_pct_change_{t.lower()}" for t in universe]
close_cols = [f"close_{t.lower()}" for t in universe]
# print("Percent columns:\n", df_universe[pct_cols].dtypes)

# Average daily close
df_universe['avg_daily_close'] = df_universe[close_cols].mean(axis=1)

def get_max_5_stocks(row):
    values = pd.to_numeric(row[pct_cols], errors='coerce') 
    if values.isna().all():  
        return None
    top_5 = values.nlargest(5).index.tolist()
    stocks = [col.replace("5day_pct_change_", '').upper() for col in top_5]
    return ', '.join(stocks)

df_universe['max_5_stocks'] = df_universe.apply(get_max_5_stocks, axis=1)
# df_universe.style.format('{:,.2f}')

def get_min_5_stocks(row):
    values = pd.to_numeric(row[pct_cols], errors='coerce')
    if values.isna().all():
        return None
    bottom_5 = values.nsmallest(5).index.tolist()
    stocks = [col.replace("5day_pct_change_",'').upper() for col in bottom_5]
    return ', '.join(stocks)

df_universe['min_5_stocks'] = df_universe.apply(get_min_5_stocks, axis=1)

df_universe = df_universe.round(2)
df_universe

,Date,adj_close_aapl,close_aapl,5day_pct_change_aapl,dividends_aapl,adj_close_amzn,close_amzn,5day_pct_change_amzn,dividends_amzn,adj_close_goog,...,close_tsla,5day_pct_change_tsla,dividends_tsla,adj_close_sap,close_sap,5day_pct_change_sap,dividends_sap,avg_daily_close,max_5_stocks,min_5_stocks
1,2018-01-02,40.30,43.06,NaN,0.0,59.45,59.45,NaN,0.0,52.82,...,21.37,NaN,0.0,99.06,112.39,NaN,0.0,77.11,None,None
2,2018-01-03,40.30,43.06,NaN,0.0,60.21,60.21,NaN,0.0,53.68,...,21.15,NaN,0.0,99.87,113.31,NaN,0.0,78.26,None,None
3,2018-01-04,40.48,43.26,NaN,0.0,60.48,60.48,NaN,0.0,53.88,...,20.97,NaN,0.0,101.40,115.05,NaN,0.0,78.89,None,None
4,2018-01-05,40.95,43.75,NaN,0.0,61.46,61.46,NaN,0.0,54.66,...,21.11,NaN,0.0,102.53,116.33,NaN,0.0,79.76,None,None
5,2018-01-08,40.79,43.59,NaN,0.0,62.34,62.34,NaN,0.0,54.90,...,22.43,NaN,0.0,101.18,114.80,NaN,0.0,80.15,None,None
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
246,2018-12-21,35.80,37.68,-8.91,0.0,68.87,68.87,-13.47,0.0,48.58,...,21.32,-12.56,0.0,86.03,96.14,-4.80,0.0,67.09,"SAP, ORCL, GOOG, MSFT, IBM","AMZN, META, TSLA, AAPL, NFLX"
247,2018-12-24,34.87,36.71,-10.44,0.0,67.20,67.20,-11.63,0.0,48.41,...,19.69,-15.22,0.0,85.01,95.00,-4.64,0.0,65.45,"GOOG, SAP, ORCL, IBM, MSFT","TSLA, AMZN, META, NFLX, AAPL"
248,2018-12-26,37.33,39.29,-5.36,0.0,73.54,73.54,-5.19,0.0,51.55,...,21.74,-3.25,0.0,88.04,98.38,-1.70,0.0,69.61,"GOOG, SAP, ORCL, TSLA, MSFT","META, NFLX, AAPL, AMZN, IBM"
249,2018-12-27,37.08,39.04,-2.95,0.0,73.08,73.08,-2.24,0.0,51.77,...,21.08,-5.06,0.0,88.59,98.99,-1.12,0.0,69.94,"GOOG, META, SAP, AMZN, IBM","TSLA, NFLX, ORCL, AAPL, MSFT"


### You start to manage 5 million dollars fund on Jan 02, 2018

You have a strategy to manage the fund.

1. On Jan 02 2018, you split the $\$5m$ into 5 $\$1m$, and use them to buy 5 stocks from the 10 stocks. For example, IBM close price was $\$154.25$. With $\$1m$, you can buy max 6482 shares with cost $\$999848.5$ with $\$151.5$‬ cash left. You decided to spend $\$1m$ on each of `['IBM', 'MSFT', 'GOOG', 'AAPL', 'AMZN']` respectively and keep the rest cash into a zero-interest cash account. On Jan 02 2018, your mark to market value (MTM) is $\$5m$ if combining all stocks value and cash. Your holdings of stocks and cach account is your portfolio.

\begin{equation*}
\ MTM^t =  cash^t + \sum_{k=1}^5 Shares_k^t \times ClosePrice_k^t \
\end{equation*}

In [7]:
# 1. 

# 01/02/2018 Close Prices:

n = 1000000

# AAPL
aapl_close = df_universe.loc[1, 'close_aapl']
aapl_shares = n //aapl_close
aapl_cost = aapl_shares * aapl_close
aapl_leftover = n - aapl_cost

print(f"AAPL Close: {aapl_close:,.2f}") # AAPL Close: 43.06
print(f"AAPL Shares purchased: {aapl_shares:,}") # AAPL Shares purchased: 23,220
print(f"AAPL Cost of shares: {aapl_cost:,.2f}") # AAPL Cost of shares: 999,969.27
print(f"AAPL money leftover: {aapl_leftover:,.2f}") # AAPL money leftover: 30.73
print("---------------------------------------------")

# AMZN: 1,000,000 / 59.45 =
amzn_close = df_universe.loc[1, 'close_amzn']
amzn_shares = n // amzn_close
amzn_cost = amzn_shares * amzn_close
amzn_leftover = n - amzn_cost

print(f"AMZN Close: {amzn_close:,.2f}") # AMZN Close: 59.45
print(f"AMZN Shares purchased: {amzn_shares:,}") # AMZN Shares purchased: 16,823
print(f"AMZN Cost of shares: {amzn_cost:,.2f}") # AMZN Cost of shares: 999,999.35
print(f"AMZN money leftover: {amzn_leftover:,.2f}") # AMZN money leftover: 0.65
print("---------------------------------------------")
    
# GOOG: 1,000,000 / 53.25 = 
goog_close = df_universe.loc[1, 'close_goog']
goog_shares = n // goog_close
goog_cost = goog_shares * goog_close
goog_leftover = n - goog_cost

print(f"GOOG Close: {goog_close:,.2f}") # GOOG Close: 53.25
print(f"GOOG Shares purchased: {goog_shares:,}") # GOOG Shares purchased: 18,776
print(f"GOOG Cost of shares: {goog_cost:,.2f}") # GOOG Cost of shares: 999,999.0
print(f"GOOG money leftover: {goog_leftover:,.2f}") # GOOG money leftover: 1.0
print("---------------------------------------------")

# IBM: 1,000,000 / 147.47 =
ibm_close = df_universe.loc[1, 'close_ibm']
ibm_shares = n // ibm_close
ibm_cost = ibm_shares * ibm_close
ibm_leftover = n - ibm_cost

print(f"IBM Close: {ibm_close:,.2f}") # IBM Close: 147.47
print(f"IBM Shares purchased: {ibm_shares:,}") # IBM Shares purchased: 6,780
print(f"IBM Cost of shares: {ibm_cost:,.2f}") # IBM Cost of shares: 999,996.6
print(f"IBM money leftover: {ibm_leftover:,.2f}") # IBM money leftover: 3.4
print("---------------------------------------------")

# MSFT: 1,000,000 / 85.95 =
msft_close = df_universe.loc[1, 'close_msft']
msft_shares = n // msft_close
msft_cost = msft_shares * msft_close
msft_leftover = n - msft_cost

print(f"MSFT Close: {msft_close:,.2f}") # MSFT Close: 85.95
print(f"MSFT Shares purchased: {msft_shares:,}") # MSFT Shares purchased: 11,638
print(f"MSFT Cost of shares: {msft_cost:,.2f}") # MSFT Cost of shares: 999,999.1
print(f"MSFT money leftover: {msft_leftover:,.2f}") # MSFT money leftover: 0.9
print("---------------------------------------------")

total_leftover = aapl_leftover + amzn_leftover + goog_leftover + ibm_leftover + msft_leftover
print(f"Total money leftover: {total_leftover:,.2f}") # Total money leftover: 36.28188781749923

mtm = aapl_cost + amzn_cost + goog_cost + ibm_cost + msft_cost + total_leftover
print(f"Total money invested: {mtm:,.2f}") # Total money invested: 5,000,000.
print("---------------------------------------------")

AAPL Close: 43.06
AAPL Shares purchased: 23,223.0
AAPL Cost of shares: 999,982.38
AAPL money leftover: 17.62
---------------------------------------------
AMZN Close: 59.45
AMZN Shares purchased: 16,820.0
AMZN Cost of shares: 999,949.00
AMZN money leftover: 51.00
---------------------------------------------
GOOG Close: 53.25
GOOG Shares purchased: 18,779.0
GOOG Cost of shares: 999,981.75
GOOG money leftover: 18.25
---------------------------------------------
IBM Close: 147.47
IBM Shares purchased: 6,781.0
IBM Cost of shares: 999,994.07
IBM money leftover: 5.93
---------------------------------------------
MSFT Close: 85.95
MSFT Shares purchased: 11,634.0
MSFT Cost of shares: 999,942.30
MSFT money leftover: 57.70
---------------------------------------------
Total money leftover: 150.50
Total money invested: 5,000,000.00
---------------------------------------------


In [8]:
# Selling all stocks on 01/09/2018:

# AAPL
aapl_sell_price = round(df_universe.loc[6, 'close_aapl'], 2)
aapl_sell_value = round(aapl_shares * aapl_sell_price, 2)
aapl_sell_profit = round(aapl_sell_value - aapl_cost, 2)

print(f"AAPL sell price on {df_universe.loc[6, 'Date']}: {aapl_sell_price:,}")
print(f"AAPL value of all shares on {df_universe.loc[6, 'Date']}: {aapl_sell_value:,}")
print(f"AAPL profit on {df_universe.loc[6, 'Date']}: {aapl_sell_profit:,}")
print("---------------------------------------------")

# AMZN
amzn_sell_price = round(df_universe.loc[6, 'close_amzn'], 2)
amzn_sell_value = round(amzn_shares * amzn_sell_price, 2)
amzn_sell_profit = round(amzn_sell_value - amzn_cost, 2)

print(f"AMZN sell price on {df_universe.loc[6, 'Date']}: {amzn_sell_price:,}")
print(f"AMZN value of all shares on {df_universe.loc[6, 'Date']}: {amzn_sell_value:,}")
print(f"AMZN profit on {df_universe.loc[6, 'Date']}: {amzn_sell_profit:,}")
print("---------------------------------------------")

# GOOG
goog_sell_price = round(df_universe.loc[6, 'close_goog'], 2)
goog_sell_value = round(goog_shares * goog_sell_price, 2)
goog_sell_profit = round(goog_sell_value - goog_cost, 2)

print(f"GOOG sell price on {df_universe.loc[6, 'Date']}: {goog_sell_price:,}")
print(f"GOOG value of all shares on {df_universe.loc[6, 'Date']}: {goog_sell_value:,}")
print(f"GOOG profit on {df_universe.loc[6, 'Date']}: {goog_sell_profit:,}")
print("---------------------------------------------")

# IBM
ibm_sell_price = round(df_universe.loc[6, 'close_ibm'], 2)
ibm_sell_value = round(ibm_shares * ibm_sell_price, 2)
ibm_sell_profit = round(ibm_sell_value - ibm_cost, 2)

print(f"IBM sell price on {df_universe.loc[6, 'Date']}: {ibm_sell_price:,}")
print(f"IBM value of all shares on {df_universe.loc[6, 'Date']}: {ibm_sell_value:,}")
print(f"IBM profit on {df_universe.loc[6, 'Date']}: {ibm_sell_profit:,}")
print("---------------------------------------------")

# MSFT
msft_sell_price = round(df_universe.loc[6, 'close_msft'], 2)
msft_sell_value = round(msft_shares * msft_sell_price, 2)
msft_sell_profit = round(msft_sell_value - msft_cost, 2)

print(f"MSFT sell price on {df_universe.loc[6, 'Date']}: {msft_sell_price:,}")
print(f"MSFT value of all shares on {df_universe.loc[6, 'Date']}: {msft_sell_value:,}")
print(f"MSFT profit on {df_universe.loc[6, 'Date']}: {msft_sell_profit:,}")
print("---------------------------------------------")

total_sell_value = aapl_sell_value + amzn_sell_value + goog_sell_value + ibm_sell_value + msft_sell_value
total_profit = aapl_sell_profit + amzn_sell_profit + goog_sell_profit + ibm_sell_profit + msft_sell_profit

print(f"Total profit on {df_universe.loc[6, 'Date']}: {total_profit:,}")
print(f"New MTM value on {df_universe.loc[6, 'Date']}: {total_sell_value:,.2f}")
print(f"New available budget per stock = {total_sell_value / 5:,.2f}")
print("---------------------------------------------")

AAPL sell price on 2018-01-09: 43.58
AAPL value of all shares on 2018-01-09: 1,012,058.34
AAPL profit on 2018-01-09: 12,075.96
---------------------------------------------
AMZN sell price on 2018-01-09: 62.63
AMZN value of all shares on 2018-01-09: 1,053,436.6
AMZN profit on 2018-01-09: 53,487.6
---------------------------------------------
GOOG sell price on 2018-01-09: 55.31
GOOG value of all shares on 2018-01-09: 1,038,666.49
GOOG profit on 2018-01-09: 38,684.74
---------------------------------------------
IBM sell price on 2018-01-09: 156.63
IBM value of all shares on 2018-01-09: 1,062,108.03
IBM profit on 2018-01-09: 62,113.96
---------------------------------------------
MSFT sell price on 2018-01-09: 88.22
MSFT value of all shares on 2018-01-09: 1,026,351.48
MSFT profit on 2018-01-09: 26,409.18
---------------------------------------------
Total profit on 2018-01-09: 192,771.43999999997
New MTM value on 2018-01-09: 5,192,620.94
New available budget per stock = 1,038,524.19
---

2. Your trading strategy is "5 days rebalancing of buying low". Here is how it works. You keep your portfolio unchanged until 5 days later on Jan 09 2018. Now you want to re-check the market and adjust your portfolio. You will compute the "Adj Close" price changes from Jan 02 to Jan 09, and find the 5 stocks whose "Adj Close" prices dropped the most in terms of percentage. You sell all current holdings on Jan 09 "Close" prices to convert your portfolio to all cash. Then immediately split your cash, including your cash account, to 5 equal parts to buy the 5 stocks that dropped the most from Jan 02 to Jan 09 on 'Adj Close' prices. You always buy the max shares of stock on the "Close" price and keep the rest cash in cash account. Now the portfolio should be different from 5 days ago. This operation is called "rebalancing".

    Keep in mind, the MTM will change every day, even when your portfolio holdings don't change, because the stock prices change.

In [9]:
# 2. Rebalancing on 01/09/2018 - Buying top 5 stocks

# New budget from selling all stocks
new_budget = 5192490.20
n2 = new_budget / 5  # Budget per stock

print(f"New MTM: ${new_budget:,.2f}")
print(f"Budget per stock: ${n2:,.2f}")
print("=============================================")

# Highest % increase: IBM, ORCL, AMZN, TSLA, NFLX

# IBM
ibm_close_2 = df_universe.loc[6, 'close_ibm']
ibm_shares_2 = int(n2 // ibm_close_2)
ibm_cost_2 = ibm_shares_2 * ibm_close_2
ibm_leftover_2 = n2 - ibm_cost_2

print(f"IBM Close: ${ibm_close_2:,.2f}")
print(f"IBM Shares purchased: {ibm_shares_2:,}")
print(f"IBM Cost of shares: ${ibm_cost_2:,.2f}")
print(f"IBM money leftover: ${ibm_leftover_2:,.2f}")
print("---------------------------------------------")

# ORCL
orcl_close_2 = df_universe.loc[6, 'close_orcl']
orcl_shares_2 = int(n2 // orcl_close_2)
orcl_cost_2 = orcl_shares_2 * orcl_close_2
orcl_leftover_2 = n2 - orcl_cost_2

print(f"ORCL Close: ${orcl_close_2:,.2f}")
print(f"ORCL Shares purchased: {orcl_shares_2:,}")
print(f"ORCL Cost of shares: ${orcl_cost_2:,.2f}")
print(f"ORCL money leftover: ${orcl_leftover_2:,.2f}")
print("---------------------------------------------")

# AMZN
amzn_close_2 = df_universe.loc[6, 'close_amzn']
amzn_shares_2 = int(n2 // amzn_close_2)
amzn_cost_2 = amzn_shares_2 * amzn_close_2
amzn_leftover_2 = n2 - amzn_cost_2

print(f"AMZN Close: ${amzn_close_2:,.2f}")
print(f"AMZN Shares purchased: {amzn_shares_2:,}")
print(f"AMZN Cost of shares: ${amzn_cost_2:,.2f}")
print(f"AMZN money leftover: ${amzn_leftover_2:,.2f}")
print("---------------------------------------------")

# TSLA
tsla_close_2 = df_universe.loc[6, 'close_tsla']
tsla_shares_2 = int(n2 // tsla_close_2)
tsla_cost_2 = tsla_shares_2 * tsla_close_2
tsla_leftover_2 = n2 - tsla_cost_2

print(f"TSLA Close: ${tsla_close_2:,.2f}")
print(f"TSLA Shares purchased: {tsla_shares_2:,}")
print(f"TSLA Cost of shares: ${tsla_cost_2:,.2f}")
print(f"TSLA money leftover: ${tsla_leftover_2:,.2f}")
print("---------------------------------------------")

# NFLX
nflx_close_2 = df_universe.loc[6, 'close_nflx']
nflx_shares_2 = int(n2 // nflx_close_2)
nflx_cost_2 = nflx_shares_2 * nflx_close_2
nflx_leftover_2 = n2 - nflx_cost_2

print(f"NFLX Close: ${nflx_close_2:,.2f}")
print(f"NFLX Shares purchased: {nflx_shares_2:,}")
print(f"NFLX Cost of shares: ${nflx_cost_2:,.2f}")
print(f"NFLX money leftover: ${nflx_leftover_2:,.2f}")
print("---------------------------------------------")

total_leftover_2 = ibm_leftover_2 + orcl_leftover_2 + amzn_leftover_2 + tsla_leftover_2 + nflx_leftover_2
total_cost_2 = ibm_cost_2 + orcl_cost_2 + amzn_cost_2 + tsla_cost_2 + nflx_cost_2
mtm_2 = total_cost_2 + total_leftover_2

print(f"Total money leftover: ${total_leftover_2:,.2f}")
print(f"Total invested: ${total_cost_2:,.2f}")
print(f"MTM on {df_universe.loc[6, 'Date']}: ${mtm_2:,.2f}")
print("---------------------------------------------")


New MTM: $5,192,490.20
Budget per stock: $1,038,498.04
IBM Close: $156.63
IBM Shares purchased: 6,630
IBM Cost of shares: $1,038,456.90
IBM money leftover: $41.14
---------------------------------------------
ORCL Close: $49.06
ORCL Shares purchased: 21,167
ORCL Cost of shares: $1,038,453.02
ORCL money leftover: $45.02
---------------------------------------------
AMZN Close: $62.63
AMZN Shares purchased: 16,581
AMZN Cost of shares: $1,038,468.03
AMZN money leftover: $30.01
---------------------------------------------
TSLA Close: $22.25
TSLA Shares purchased: 46,674
TSLA Cost of shares: $1,038,496.50
TSLA money leftover: $1.54
---------------------------------------------
NFLX Close: $20.93
NFLX Shares purchased: 49,617
NFLX Cost of shares: $1,038,483.81
NFLX money leftover: $14.23
---------------------------------------------
Total money leftover: $131.94
Total invested: $5,192,358.26
MTM on 2018-01-09: $5,192,490.20
---------------------------------------------


In [10]:
# Iitial Investment: $5,000,000.00

# AAPL Close: 43.06
# AAPL Shares purchased: 23,220.0
# AAPL Cost of shares: 999,969.27
# AAPL money leftover: 30.73
# ---------------------------------------------
# AMZN Close: 59.45
# AMZN Shares purchased: 16,820.0
# AMZN Cost of shares: 999,957.42
# AMZN money leftover: 42.58
# ---------------------------------------------
# GOOG Close: 53.25
# GOOG Shares purchased: 18,779.0
# GOOG Cost of shares: 999,981.75
# GOOG money leftover: 18.25
# ---------------------------------------------
# IBM Close: 147.47
# IBM Shares purchased: 6,781.0
# IBM Cost of shares: 999,970.59
# IBM money leftover: 29.41
# ---------------------------------------------
# MSFT Close: 85.95
# MSFT Shares purchased: 11,634.0
# MSFT Cost of shares: 999,942.26
# MSFT money leftover: 57.74
# ---------------------------------------------
# Total money leftover: 178.71
# Total money invested: 5,000,000.00
# ---------------------------------------------


In [11]:

# First rebalanace: 01/09/2018

# AAPL sell price on 2018-01-09: 43.58
# AAPL value of all shares on 2018-01-09: 1,011,927.6
# AAPL profit on 2018-01-09: 11,958.33
# ---------------------------------------------
# AMZN sell price on 2018-01-09: 62.63
# AMZN value of all shares on 2018-01-09: 1,053,436.6
# AMZN profit on 2018-01-09: 53,479.18
# ---------------------------------------------
# GOOG sell price on 2018-01-09: 55.31
# GOOG value of all shares on 2018-01-09: 1,038,666.49
# GOOG profit on 2018-01-09: 38,684.74
# ---------------------------------------------
# IBM sell price on 2018-01-09: 156.63
# IBM value of all shares on 2018-01-09: 1,062,108.03
# IBM profit on 2018-01-09: 62,137.44
# ---------------------------------------------
# MSFT sell price on 2018-01-09: 88.22
# MSFT value of all shares on 2018-01-09: 1,026,351.48
# MSFT profit on 2018-01-09: 26,409.22
# ---------------------------------------------
# Total profit on 2018-01-09: 192,668.91
# New MTM value on 2018-01-09: 5,192,490.20
# New available budget per stock = 1,038,498.04
# ---------------------------------------------

In [12]:

# 2nd Investment: 01/09/2018

# New MTM value on 2018-01-09: 5,192,490.20
# New available budget per stock = 1,038,498.04

# IBM Close: $156.63
# IBM Shares purchased: 6,630
# IBM Cost of shares: $1,038,425.37
# IBM money leftover: $72.67
# ---------------------------------------------
# ORCL Close: $49.06
# ORCL Shares purchased: 21,167
# ORCL Cost of shares: $1,038,453.05
# ORCL money leftover: $44.99
# ---------------------------------------------
# AMZN Close: $62.63
# AMZN Shares purchased: 16,580
# AMZN Cost of shares: $1,038,488.27
# AMZN money leftover: $9.77
# ---------------------------------------------
# TSLA Close: $22.25
# TSLA Shares purchased: 46,682
# TSLA Cost of shares: $1,038,487.79
# TSLA money leftover: $10.25
# ---------------------------------------------
# NFLX Close: $20.93
# NFLX Shares purchased: 49,615
# NFLX Cost of shares: $1,038,491.55
# NFLX money leftover: $6.49
# ---------------------------------------------
# Total money leftover: $144.17
# Total invested: $5,192,346.03
# MTM on 2018-01-09: $5,192,490.20


3. Corporations generally issue stock dividends on some days. The total dividend you get on such a day is the stock dividend  times your shares if you have shares of this stock on the dividend day. If you buy shares on the dividend day, these bought shares are not qualified to get dividend. If you sell shares on the dividend day, the sold shares are qualified to get dividend. For example, on 2/8/2018, IBM issued \$1.5 dividend per share. In your cash account, you will automatically get

\begin{equation*}
\$1.5 \times \left(your\ IBM\ shares\ on\ 2/8/2018\right)
\end{equation*} 

In [13]:
# 3. 



4. 5 business days later on Jan 17 (Jan 15 was a holiday), you re-check the market and adjust your portfolio again. You will have a new portfolio on Jan 17.

In [14]:
# 4. 

5. If you run this strategy every 5 days all the way to Dec 31 2018, you will have a daily MTM. You expect the MTM on Dec 31 2018 should be higher than \$5m because you always buy the stocks that dropped the most, i.e., you always buy low.

In [15]:
# 5. 

6. Another strategy is "5 days rebalancing of buying high". You always buy the 5 stocks whose "Adj Close" prices surge the most in terms of percentage because you believe the trend will continue. Run the new strategy and see how the MTM will change.

In [16]:
# 9. 

rebalance_days = list(range(1, len(df_universe) + 1, 5)) 
# print(f"Rebalancing days 2018: \n {rebalance_days}\n")
# print(f"Total trading days 2018: {len(df_universe)}")
# print("----------------------------------------")

interval = 5
i = 250

# ========================= Buying High ======================== #

initial_stocks = ['ibm','msft','goog','aapl','amzn']
budget_per_stock = 1_000_000
holdings = {}  
holdings_history = []
cash = 0
mtm_list_high = []  

for t in initial_stocks:
    close = df_universe.loc[1, f"close_{t}"]
    shares = int(budget_per_stock // close)
    cost = shares * close
    holdings[t] = shares
    cash += (budget_per_stock - cost)

# print(f"Initial holdings: {holdings}")
# print(f"Initial leftover cash: {cash:,.2f}")
# print("----------------------------------------")

# -----------------------------------------------

for day in range(1, len(df_universe) + 1):

    for t in holdings:
        div_col = f'dividends_{t}'
        dividend = df_universe.loc[day, div_col]
        if dividend > 0:
            cash += holdings[t] * dividend

    if day > 1 and (day -1 ) % interval == 0:
        max_5 = df_universe.loc[day, 'max_5_stocks']
        if max_5 is not None:
            for t in holdings:
                cash += holdings[t] * df_universe.loc[day,f'close_{t}']
            holdings = {}
            share_budget = cash / 5

            max_5_stocks = max_5.split(', ')
            for stock in max_5_stocks:
                stock = stock.lower()
                close = df_universe.loc[day, f'close_{stock}']
                shares = int(share_budget // close)
                cost = shares * close
                holdings[stock] = shares
                cash -= cost
    
    holdings_history.append(dict(holdings)) 

    stock_value = sum(holdings[t] * df_universe.loc[day, f'close_{t}'] for t in holdings)
    mtm = cash + stock_value
    mtm_list_high.append(mtm)

# --------------------------------------------------------

print(f"Buying High Strategy rebalancing every {interval} days: \n")
print(f"Holdings on {df_universe.loc[i, 'Date']}: \n{holdings_history[i-1]} \n")
print(f"MTM on day 1: ${mtm_list_high[0]:,.2f}")
print(f"MTM on day {i}: ${mtm_list_high[i-1]:,.2f}")
print(f"Net Gain/Loss: ${mtm_list_high[i-1] - mtm_list_high[0]:,.2f} \n")
print("-" * 90)

# ========================= Buying Low ======================== #

initial_stocks = ['ibm','msft','goog','aapl','amzn']
budget_per_stock = 1_000_000
holdings = {}  
holdings_history = []
cash = 0
mtm_list_low = []  

for t in initial_stocks:
    close = df_universe.loc[1, f"close_{t}"]
    shares = int(budget_per_stock // close)
    cost = shares * close
    holdings[t] = shares
    cash += (budget_per_stock - cost)

# print(f"Initial holdings: {holdings}")
# print(f"Initial leftover cash: {cash:,.2f}")
# print("----------------------------------------")

# -----------------------------------------------------------

for day in range(1, len(df_universe) + 1):

    for t in holdings:
        div_col = f'dividends_{t}'
        dividend = df_universe.loc[day, div_col]
        if dividend > 0:
            cash += holdings[t] * dividend

    if day > 1 and (day -1 ) % interval == 0:
        min_5 = df_universe.loc[day, 'min_5_stocks']
        if min_5 is not None:
            for t in holdings:
                cash += holdings[t] * df_universe.loc[day,f'close_{t}']
            holdings = {}
            share_budget = cash / 5

            min_5_stocks = min_5.split(', ')
            for stock in min_5_stocks:
                stock = stock.lower()
                close = df_universe.loc[day, f'close_{stock}']
                shares = int(share_budget // close)
                cost = shares * close
                holdings[stock] = shares
                cash -= cost
    
    holdings_history.append(dict(holdings)) 

    stock_value = sum(holdings[t] * df_universe.loc[day, f'close_{t}'] for t in holdings)
    mtm = cash + stock_value
    mtm_list_low.append(mtm)

# ---------------------------------------------------------------

print(f"Buying Low Strategy rebalancing every {interval} days: \n")
print(f"Holdings on {df_universe.loc[i, 'Date']}: \n{holdings_history[i-1]} \n")
print(f"MTM on day 1: ${mtm_list_low[0]:,.2f}")
print(f"MTM on day {i}: ${mtm_list_low[i-1]:,.2f}")
print(f"Net Gain/Loss: ${mtm_list_low[i-1] - mtm_list_low[0]:,.2f} \n")

Buying High Strategy rebalancing every 5 days: 

Holdings on 2018-12-28: 
{'sap': 9758, 'orcl': 21321, 'goog': 19154, 'msft': 9550, 'ibm': 8845} 

MTM on day 1: $5,000,000.00
MTM on day 250: $4,831,471.23
Net Gain/Loss: $-168,528.77 

------------------------------------------------------------------------------------------
Buying Low Strategy rebalancing every 5 days: 

Holdings on 2018-12-28: 
{'amzn': 14691, 'meta': 8097, 'tsla': 47457, 'aapl': 26852, 'nflx': 41062} 

MTM on day 1: $5,000,000.00
MTM on day 250: $5,321,148.94
Net Gain/Loss: $321,148.94 



In [17]:
# 7.

fig = go.Figure()

fig.add_trace(
    go.Scatter(
        y=mtm_list_high,
        mode='lines',
        name='Buying High',
    )
)

fig.add_trace(
        go.Scatter(
        y=mtm_list_low,
        mode='lines',
        name='Buying Low',
    )
)

fig.update_layout(
    title='MTM: Buying High vs Buying Low Strategy',
    title_x=0.5,
    xaxis_title='Trading Days',
    yaxis_title='MTM Value ($)',
    legend_title='Strategy',
    hovermode='x unified',
)

fig.show()

7. You will create a "high tech index" which is simply the daily average of the 10 stocks "Close" prices. Compare your MTM series with the "high tech index" and plot their curves. To plot the two curves together, you may want to convert the series to daily percentage change with regard to Jan 02 2018.

In [18]:
# 7. calculate daily average of the 10 stocks close price:

high_tech_index = df_universe['avg_daily_close'].values
high_tech_pct = [(price - high_tech_index[0]) / high_tech_index[0] * 100 for price in high_tech_index]

# high_tech_index
high_tech_pct

# mtm_high_pct = [(mtm - mtm_list_high[0]) / mtm_list_high[0] * 100 for mtm in mtm_list_high]
# mtm_low_pct = [(mtm - mtm_list_low[0]) / mtm_list_low[0] * 100 for mtm in mtm_list_low]

[np.float64(0.0),
 np.float64(1.4913759564258924),
 np.float64(2.308390610815719),
 np.float64(3.436648943068351),
 np.float64(3.942419919595391),
 np.float64(3.8257035403968396),
 np.float64(3.657113214887832),
 np.float64(3.566333808844508),
 np.float64(3.073531318895091),
 np.float64(2.956814939696539),
 np.float64(3.942419919595391),
 np.float64(4.331474516923879),
 np.float64(3.9942938659058465),
 np.float64(5.109583711580855),
 np.float64(6.990014265335236),
 np.float64(6.484243288808196),
 np.float64(6.613928154584353),
 np.float64(8.079367137855018),
 np.float64(7.417974322396575),
 np.float64(6.445337829075346),
 np.float64(7.223447023732341),
 np.float64(7.054856698223315),
 np.float64(5.083646738425628),
 np.float64(0.9077940604331511),
 np.float64(3.5793022954221305),
 np.float64(1.5951238490468214),
 np.float64(-2.917909479963688),
 np.float64(-1.3616910906497175),
 np.float64(-0.06484243288807827),
 np.float64(-0.11671637919855196),
 np.float64(2.295422124238096),
 np.flo

In [19]:

fig1 = go.Figure()

fig1.add_trace(
    go.Scatter(
        y=high_tech_index,
        mode='lines',
        name='daily avg. close'
    )
)

fig1.update_layout(
    title='Daily Average Close Price of 10 Stocks',
    title_x=0.5,
    xaxis_title='Trading Days',
    yaxis_title='Price ($)',
    legend_title='Stock',
    hovermode='x unified',
)


8. Download the USD/JPY 2018 historical data at https://finance.yahoo.com/quote/JPY%3DX/history?period1=1514764800&period2=1546300800&interval=1d&filter=history&frequency=1d&includeAdjustedClose=true then use the "Close" column as the rate to convert your MTM series from USD to JPY. Plot the two MTM curves. You will need to convert to daily percentage change too.

In [20]:
# 8.

# 2018 USD --> JPY: USD x 109

mtm_low_jpy = [price * 109 for price in mtm_list_low]
mtm_high_jpy = [price * 109 for price in mtm_list_high]

mtm_low_jpy

[np.float64(545000000.0),
 np.float64(551674896.22),
 np.float64(556318439.01),
 np.float64(562896452.76),
 np.float64(565404987.48),
 np.float64(566012086.9599999),
 np.float64(563807262.2099999),
 np.float64(561852276.36),
 np.float64(562094621.51),
 np.float64(559833175.62),
 np.float64(565099607.63),
 np.float64(566627721.1400001),
 np.float64(570571251.7600001),
 np.float64(576868446.63),
 np.float64(583618622.61),
 np.float64(582001063.6999999),
 np.float64(578192686.54),
 np.float64(586339069.6800001),
 np.float64(583789099.7),
 np.float64(575908240.56),
 np.float64(583467826.5600001),
 np.float64(585079199.72),
 np.float64(569881767.9000001),
 np.float64(548831225.74),
 np.float64(563465868.7600001),
 np.float64(552175520.1400001),
 np.float64(529949490.37000006),
 np.float64(541400298.98),
 np.float64(547173095.9400002),
 np.float64(548391081.5600001),
 np.float64(559484094.4000001),
 np.float64(572985429.5400001),
 np.float64(570584404.7900002),
 np.float64(566914155.7),
 np.

In [21]:
# 8.

fig2 = go.Figure()

fig2.add_trace(
    go.Scatter(
        y=mtm_low_jpy,
        mode='lines',
        name='MTM Low (JPY)'
    )
)

fig2.add_trace(
    go.Scatter(
        y=mtm_high_jpy,
        mode='lines',
        name='MTM High (JPY)'
    )
)

fig2.update_layout(
    title='MTM High vs Low (JPY)',
    title_x=0.5,
    xaxis_title='Trading Days',
    yaxis_title='Price ($)',
    legend_title='Stock',
    hovermode='x unified',
)

9. The above two strategies both rebalance every 5 days. Try to change the days interval and find the optimal days interval that maximizes the MTM on 12/31/2018. 

In [38]:
# 9. 

rebalance_days = list(range(1, len(df_universe) + 1, 5)) 
# print(f"Rebalancing days 2018: \n {rebalance_days}\n")
# print(f"Total trading days 2018: {len(df_universe)}")
# print("----------------------------------------")

interval = 5
i = 250

# ========================= Buying High ======================== #

initial_stocks = ['ibm','msft','goog','aapl','amzn']
budget_per_stock = 1_000_000
holdings = {}  
holdings_history = []
cash = 0
mtm_list_high = []  

for t in initial_stocks:
    close = df_universe.loc[1, f"close_{t}"]
    shares = int(budget_per_stock // close)
    cost = shares * close
    holdings[t] = shares
    cash += (budget_per_stock - cost)

# print(f"Initial holdings: {holdings}")
# print(f"Initial leftover cash: {cash:,.2f}")
# print("----------------------------------------")

# -----------------------------------------------

for day in range(1, len(df_universe) + 1):

    for t in holdings:
        div_col = f'dividends_{t}'
        dividend = df_universe.loc[day, div_col]
        if dividend > 0:
            cash += holdings[t] * dividend

    if day > 1 and (day -1 ) % interval == 0:
        max_5 = df_universe.loc[day, 'max_5_stocks']
        if max_5 is not None:
            for t in holdings:
                cash += holdings[t] * df_universe.loc[day,f'close_{t}']
            holdings = {}
            share_budget = cash / 5

            max_5_stocks = max_5.split(', ')
            for stock in max_5_stocks:
                stock = stock.lower()
                close = df_universe.loc[day, f'close_{stock}']
                shares = int(share_budget // close)
                cost = shares * close
                holdings[stock] = shares
                cash -= cost
    
    holdings_history.append(dict(holdings)) 

    stock_value = sum(holdings[t] * df_universe.loc[day, f'close_{t}'] for t in holdings)
    mtm = cash + stock_value
    mtm_list_high.append(mtm)

# --------------------------------------------------------

# print(Fore.YELLOW + f"Buying High Strategy rebalancing every {interval} days: \n" + Style.RESET_ALL)
print(f"Buying High Strategy rebalancing every " + Fore.YELLOW + Style.BRIGHT + f"{interval}" + Style.RESET_ALL + " days: \n")
print(Fore.GREEN + f"Holdings on {df_universe.loc[i, 'Date']}: \n{holdings_history[i-1]} \n" + Style.RESET_ALL)
print(f"MTM on day 1: ${mtm_list_high[0]:,.2f}" + Style.RESET_ALL)
print(f"MTM on day {i}: ${mtm_list_high[i-1]:,.2f}" + Style.RESET_ALL)
print(Fore.RED + f"Net Gain/Loss: ${mtm_list_high[i-1] - mtm_list_high[0]:,.2f} \n" + Style.RESET_ALL)
print("-" * 90)

# ========================= Buying Low ======================== #

initial_stocks = ['ibm','msft','goog','aapl','amzn']
budget_per_stock = 1_000_000
holdings = {}  
holdings_history = []
cash = 0
mtm_list_low = []  

for t in initial_stocks:
    close = df_universe.loc[1, f"close_{t}"]
    shares = int(budget_per_stock // close)
    cost = shares * close
    holdings[t] = shares
    cash += (budget_per_stock - cost)

# print(f"Initial holdings: {holdings}")
# print(f"Initial leftover cash: {cash:,.2f}")
# print("----------------------------------------")

# -----------------------------------------------------------

for day in range(1, len(df_universe) + 1):

    for t in holdings:
        div_col = f'dividends_{t}'
        dividend = df_universe.loc[day, div_col]
        if dividend > 0:
            cash += holdings[t] * dividend

    if day > 1 and (day -1 ) % interval == 0:
        min_5 = df_universe.loc[day, 'min_5_stocks']
        if min_5 is not None:
            for t in holdings:
                cash += holdings[t] * df_universe.loc[day,f'close_{t}']
            holdings = {}
            share_budget = cash / 5

            min_5_stocks = min_5.split(', ')
            for stock in min_5_stocks:
                stock = stock.lower()
                close = df_universe.loc[day, f'close_{stock}']
                shares = int(share_budget // close)
                cost = shares * close
                holdings[stock] = shares
                cash -= cost
    
    holdings_history.append(dict(holdings)) 

    stock_value = sum(holdings[t] * df_universe.loc[day, f'close_{t}'] for t in holdings)
    mtm = cash + stock_value
    mtm_list_low.append(mtm)

# ---------------------------------------------------------------

# print(Fore.YELLOW + f"Buying Low Strategy rebalancing every {interval} days: \n" + Style.RESET_ALL)
print(f"Buying Low Strategy rebalancing every " + Fore.YELLOW + Style.BRIGHT + f"{interval}" + Style.RESET_ALL + " days: \n")
print(Fore.GREEN + f"Holdings on {df_universe.loc[i, 'Date']}: \n{holdings_history[i-1]} \n" + Style.RESET_ALL)
print(f"MTM on day 1: ${mtm_list_low[0]:,.2f}" + Style.RESET_ALL)
print(f"MTM on day {i}: ${mtm_list_low[i-1]:,.2f}" + Style.RESET_ALL)
print(Fore.BLUE + f"Net Gain/Loss: ${mtm_list_low[i-1] - mtm_list_low[0]:,.2f} \n" + Style.RESET_ALL)

Buying High Strategy rebalancing every 5 days: 

Holdings on 2018-12-28: 
{'sap': 9758, 'orcl': 21321, 'goog': 19154, 'msft': 9550, 'ibm': 8845} 

MTM on day 1: $5,000,000.00
MTM on day 250: $4,831,471.23
Net Gain/Loss: $-168,528.77 

------------------------------------------------------------------------------------------
Buying Low Strategy rebalancing every 5 days: 

Holdings on 2018-12-28: 
{'amzn': 14691, 'meta': 8097, 'tsla': 47457, 'aapl': 26852, 'nflx': 41062} 

MTM on day 1: $5,000,000.00
MTM on day 250: $5,321,148.94
Net Gain/Loss: $321,148.94 



### Submit(before midterm exam)

Submit your jupyter notebook HTML file with all intermediate results and the final dataframe. Plot the curves.

This is a group project. Each group may have up to 3 members. Solo is good too.